src/messaging/ExchangeHolder.h

```diff
@@ -0,0 +1,136 @@
+ *      by application and protocol logic code that would otherwise need to closely
+ *      manange their internal pointers to an ExchangeContext and correctly
+ *      null-it out/abort it depending on the circumstances. This relies on clear rules
+ *      established by ExchangeContext and the transfer of ownership at various points
+ *      in its lifetime.
+ *
+ *      It does this by listening in on OnExchangeClosing and looking at the various
+ *      states the exchange might be in to decide amongst a couple of different tactics.
+ *
+ *      This is a delegate forwarder - consumers can still register to be an ExchangeDelegate
+ *      and get notified of all relevant happenings on that delegate interface.
+ *
+ */
+class ExchangeHolder : public ExchangeDelegate
+{
+public:
+    ExchangeHolder(ExchangeDelegate & delegate) : mpExchangeDelegate(delegate) {}
+
+    virtual ~ExchangeHolder() { AbortIfNeeded(); }
+
+    bool Contains(const ExchangeContext * exchange) const { return mpExchangeCtx == exchange; }
+
+    /*
+     *  Aborts any previously tracked exchange and
+     *  re-acquires the provided exchange. As part of that,
+     *  this will set itself as the delegate on that exchange (and forward
+     *  calls to the registered delegate on this object).
+     *
+     */
+    void Grab(ExchangeContext * exchange)
+    {
+        AbortIfNeeded();
+
+        mpExchangeCtx = exchange;
+        mpExchangeCtx->SetDelegate(this);
+    }
+
+    /*
+     *  Releases any previously tracked exchanges by nulling
+     *  out ourselves as a delegate and then, our reference to that
+     *  exchange
+     */
+    void Release()
+    {
+        if (mpExchangeCtx)
+        {
+            mpExchangeCtx->SetDelegate(nullptr);
+        }
+
+        mpExchangeCtx = nullptr;
+    }
+
+    explicit operator bool() const { return mpExchangeCtx != nullptr; }
+    ExchangeContext * Get() const { return mpExchangeCtx; }
+
+    ExchangeContext * operator->() const
+    {
+        VerifyOrDie(mpExchangeCtx != nullptr);
+        return mpExchangeCtx;
+    }
+
+private:
+    /*
+     * This asseses if an Abort() needs to be called on the exchange. This method is only called when the holder itself
+     * is being destroyed and it's still holding onto to a reference to an exchange. It then calls Abort() only if:
+     *      1. The exchange is currently awaiting a response. This would have happened if our consumer just sent a message on the
+     * exchange and is awaiting a response. Consequently, we should go ahead and Abort() it given destruction of the holder
+     * indicates a lack of further interest in this exchange.
+     *
+     *      2. Our consumer has signaled an interest in sending a message. This could have been signaled right at exchange creation
+     * time as the initiator, or when handling a message and the consumer intends to send a response, albeit, asynchronously.
+     *
+     */
```

Comment:
The logic of whether Abort is called or not seems periphery to the primary purpose of this method (which should be stated early in the header comment):

Dereference the current ExchangeContext and associated delegate if one exists.

Reply:
There are many ways to 'dereference' an ExchangeContext, so it's important to be a bit more detailed here, since the crux of this method is specifically about the conditions in which it will call Abort, which is a specific type of de-referencing.

Reply:
I've re-factored this a bit and moved the detailed comment into the body of the function.